abra_core/src/typechecker/types.rs

```diff
@@ -32,7 +33,7 @@ pub struct StructType {
 }
 
 impl Type {
-    pub fn is_equivalent_to(&self, target_type: &Type) -> bool {
+    pub fn is_equivalent_to(&self, target_type: &Type, referencable_types: &HashMap<String, Type>) -> bool {
```

Comment:
When testing equivalence of types now, we need to have knowledge of all referencable types in order to resolve instances of `Type::Reference`. These type values could be deeply nested in a type data structure (ie. `Type::Option(Box::new(Type::Array(Box::new(Type::Reference("MyType")))))`). So we need to pass this down as we recurse.